README.md
mrepo Cookbook
==============
TODO: Enter the cookbook description here.

e.g.
This cookbook makes your favorite breakfast sandwhich.

Requirements
------------
TODO: List your cookbook requirements. Be sure to include any requirements this cookbook has on platforms, libraries, other cookbooks, packages, operating systems, etc.

e.g.
#### packages
- `toaster` - mrepo needs toaster to brown your bagel.

Attributes
----------
TODO: List you cookbook attributes here.

e.g.
#### mrepo::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['mrepo']['bacon']</tt></td>
    <td>Boolean</td>
    <td>whether to include bacon</td>
    <td><tt>true</tt></td>
  </tr>
</table>

Usage
-----
#### mrepo::default
TODO: Write usage instructions for each cookbook.

e.g.
Just include `mrepo` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[mrepo]"
  ]
}
```

Contributing
------------
TODO: (optional) If this is a public cookbook, detail the process for contributing. If this is a private cookbook, remove this section.

e.g.
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: TODO: List authors


CHANGELOG.md
# CHANGELOG for mrepo

This file is used to list changes made in each version of mrepo.

## 0.1.0:

* Initial release of mrepo

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/default.rb
#
# Cookbook Name:: mrepo
# Recipe:: default
# Author:: Jeremy MAURO <j.mauro@criteo.com>
# Description:: Recipe to setup mrepo
#
# Copyright 2013, Criteo
#
# All rights reserved - Do Not Redistribute
#
#

# --[ First of all: make sure epel is delcared]--
include_recipe 'yum::epel'

# --[ Install pkg ]--
node[:mrepo][:packages].each do |pkg|
  package pkg do
    action :install
  end
end 

template '/etc/mrepo.conf' do
  source 'mrepo.conf.erb'
  owner 'root'
  group 'root'
  mode  '0644'
  variables(
    :mrepo => node[:mrepo],
  )
end

template '/etc/logrotate.d/mrepo' do
  source 'mrepo.logrotate.erb'
  owner 'root'
  group 'root'
  mode  '0644'
  variables(
    :mrepo => node[:mrepo],
  )
end

# --[ Make directory are present ]--
dir_create = [
  "#{node[:mrepo][:keydir]}",
  "#{node[:mrepo][:srcdir]}",
  "#{node[:mrepo][:wwwdir]}",
  "#{node[:mrepo][:lockdir]}",
  "#{node[:mrepo][:cachedir]}",
  "#{node[:mrepo][:config_dir]}",
  ]
dir_create.each do |dir|
  directory dir do
    owner 'root'
    group 'root'
    mode  '0755'
    recursive true

    action :create
  end
end


definitions/mirror_repo.rb
#
# Cookbook Name::	mirror_repo.rb
# Description::		Definition to add remote repository to mrepo
# Recipe::				mrepo
# Author::        Jeremy MAURO (j.mauro@criteo.com)
#
#

# Example of use:
# mirror_repo 'CentOS-6' do
#   description 'Repository CentOS 5.6 32 bit'
#   release   '5.6'
#   arch      'i386'
#   metadata  [ 'repomd' , 'yum' ]
#   action    'create'
#   update    'nightly'
#   urls      {
#     :addons      => 'rsync://mirrors.kernel.org/centos/$release/addons/$arch/',
#     :centosplus  => 'rsync://mirrors.kernel.org/centos/$release/centosplus/$arch/',
#     :updates     => 'rsync://mirrors.kernel.org/centos/$release/updates/$arch/',
#   }
define  :mirror_repo,
        :description => nil,
        :release     => nil,
        :arch        => nil,
        :metadata    => nil,
        :key_url     => nil,
        :iso         => nil,
        :urls        => nil,
        :hour        => '0',
        :action      => 'create',
        :update      => 'weekly',
        :gentimeout  => '3600' do

  include_recipe 'mrepo'

  acceptable_action   = [ 'create', 'delete' ]
  acceptable_metadata = [ 'yum', 'apt', 'repomd', 'repoview' ]
  acceptable_update   = [ 'now', 'weekly', 'nightly', 'never' ]
  acceptable_arch     = [ 'i386', 'i586', 'x86_64', 'ppc', 's390', 's390x', 'ia64' ]


  # --[ Get the parameters ]--
  release        = params[:release]
  iso            = params[:iso]
  key_url        = params[:key_url]
  urls           = params[:urls]
  mirror_name    = params[:name]
  create         = params[:action]
  update         = params[:update]
  description    = params[:description]
  gentimeout     = params[:gentimeout].to_i
  src_dir        = node[:mrepo][:srcdir]
  key_repo       = node[:mrepo][:keydir]
  www_dir        = "#{node[:mrepo][:wwwdir]}/#{mirror_name}"
  mrepo_dir_conf = "#{node[:mrepo][:config_dir]}/#{mirror_name}.conf"
  metadata       = if params[:metadata].is_a? Array; then params[:metadata].join(' '); else params[:metadata]; end

  # --[ Meta could be an array ]--
  if params[:metadata].is_a? Array
    metadata   = params[:metadata].join(' ')
    array_meta = params[:metadata]
  else
    metadata   = params[:metadata]
    array_meta = [ metadata ]
  end

  # --[ Arch could also be an array ]--
  if params[:arch].is_a? Array
    arch       = params[:arch].join(' ')
    array_arch = params[:arch]
  else
    arch       = params[:arch]
    array_arch = [ arch ]
  end
  # --[ Check arguments ]--
  invalide_array = {
    :invalide_action => {
      :title => 'action',
      :value => [ create ] - acceptable_action,
    },
    :invalide_update => {
      :title => 'update',
      :value => [ update ] - acceptable_update,
    },
    :invalide_metadata => {
      :title => 'metadata',
      :value => array_meta - acceptable_metadata,
    },
    :invalide_arch => {
      :title => 'arch',
      :value => array_arch - acceptable_arch,
    },
  }

  invalide_options = %w(invalide_action invalide_metadata invalide_update invalide_arch)
  invalide_options.each do |option|
    if invalide_array[:"#{option}"][:value].size == 1
      title = [:"#{option}"][:title]
      value = [:"#{option}"][:value]
      Chef::Application.fatal! "The passed value [#{value}] for the option [#{title}] is not an acceptable value"
    end
  end

  unless key_url.nil?
    remote_file "#{key_repo}/#{mirror_name}" do
      owner 'root'
      group 'root'
      mode '0644'
      source key_url
    end
  end

  if create == 'create'
    Chef::Log.info ">>> [:mirror_repo] Adding repo '#{mirror_name}'"
    template mrepo_dir_conf do
      source 'repo.conf.erb'
      owner 'root'
      group 'root'
      mode  '0644'
      variables(
        :name        => mirror_name,
        :description => description,
        :release     => release,
        :arch        => arch,
        :metadata    => metadata,
        :iso         => iso,
        :urls        => urls,
      )
      notifies :run, "execute[Generate mrepo for #{mirror_name}]"
    end

    Chef::Log.info ">>> [:mirror_repo] Generating repo '#{mirror_name}'"
    execute "Generate mrepo for #{mirror_name}" do
      path ['/usr/bin','/bin']
      command "mrepo -g #{mirror_name}"
      cwd src_dir
      user 'root'
      group 'root'
      timeout gentimeout
    
      action :nothing
    end

    directory www_dir do
      owner "root"
      group "root"
      mode "0755"
    
      action :create
    end

    if params[:update] =~ /(?i-mx:now)/
      # --[ Update repo is now ]--
      Chef::Log.info ">>> [:mirror_repo] Synchronizing now repo '#{mirror_name}'"
      execute "Synchronize repo #{mirror_name}" do
        path ['/usr/bin','/bin']
        command "/usr/bin/mrepo -gu #{mirror_name}"
        cwd src_dir
        user "root"
        group "root"
        timeout gentimeout
      
        action :run
      end

      # --[ Removing Crons ]--
      cron "Nightly synchronize repo #{mirror_name}" do

        action :delete
      end
      
      cron "Weekly synchronize repo #{mirror_name}" do

        action :delete
      end

    elsif params[:update] =~ /(?i-mx:nightly)/
      Chef::Log.info ">>> [:mirror_repo] Setting nightly cron for '#{mirror_name}'"
      # --[ Update repo is done every night ]--
      cron "Nightly synchronize repo #{mirror_name}" do
        hour params[:hour]
        minute '0'
        path "/bin:/usr/bin"
        command "/usr/bin/mrepo -gu #{mirror_name}"
        user "root"
        home src_dir
        shell "/bin/bash"

        action :create
      end

      cron "Weekly synchronize repo #{mirror_name}" do

        action :delete
      end

    elsif params[:update] =~ /(?i-mx:weekly)/
      Chef::Log.info ">>> [:mirror_repo] Setting weekly cron for '#{mirror_name}'"
      # --[ Update repo is done every week ]--
      cron "Weekly synchronize repo #{mirror_name}" do
        weekday '0'
        hour params[:hour]
        minute '0'
        path "/bin:/usr/bin"
        command "/usr/bin/mrepo -gu #{mirror_name}"
        user "root"
        home src_dir
        shell "/bin/bash"

        action :create
      end
      
      cron "Nightly synchronize repo #{mirror_name}" do

        action :delete
      end
    end
  else
    Chef::Log.info ">>> [:mirror_repo] Removing repo '#{mirror_name}'"
    Chef::Log.info ">>> [:mirror_repo] Umounting iso for repo '#{mirror_name}'"
    execute "Unmount any mirrored ISOs for #{mirror_name}" do
      path ['/usr/bin', '/bin', '/usr/sbin', '/sbin']
      command "umount #{www_dir}/disc*"
      user "root"
      group "root"
      if_only "/bin/mount | /bin/grep #{www_dir}/disk"
    end

    Chef::Log.info ">>> [:mirror_repo] Removing files for repo '#{mirror_name}'"
    dir_to_remove = %W(www_dir mrepo_dir_conf)
    dir_to_remove.each do |dir|
      directory dir do
        recursive true
      
        action :delete
      end
    end

    Chef::Log.info ">>> [:mirror_repo] Removing any remaining cron for '#{mirror_name}'"
    cron "Nightly synchronize repo #{mirror_name}" do

      action :delete
    end
  end
end
